File: src/kotlin/rules_intellij/indexing/OnDiskSharedIndexChunkLocator.java
// Copyright 2000-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package rules_intellij.indexing;

import com.intellij.indexing.shared.message.SharedIndexesBundle;
import com.intellij.indexing.shared.metadata.SharedIndexMetadata;
import com.intellij.indexing.shared.platform.api.ChunkDescriptor;
import com.intellij.indexing.shared.platform.api.SharedIndexInfrastructureVersion;
import com.intellij.indexing.shared.platform.impl.ChunkStorageOption;
import com.intellij.indexing.shared.platform.impl.SharedIndexChunkConfiguration;
import com.intellij.indexing.shared.platform.impl.SharedIndexesFusCollector;
import com.intellij.indexing.shared.util.zipFs.UncompressedZipFileSystem;
import com.intellij.indexing.shared.util.zipFs.UncompressedZipFileSystemProvider;
import com.intellij.openapi.application.PathManager;
import com.intellij.openapi.diagnostic.ControlFlowException;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.extensions.ExtensionPointName;
import com.intellij.openapi.progress.ProgressIndicator;
import com.intellij.openapi.project.DumbModeTask;
import com.intellij.openapi.project.DumbService;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.startup.StartupActivity;
import com.intellij.openapi.util.io.FileUtil;
import com.intellij.openapi.util.text.StringUtil;
import com.intellij.util.ExceptionUtil;
import com.intellij.util.containers.ContainerUtil;
import com.intellij.util.io.PathKt;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Collections;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class OnDiskSharedIndexChunkLocator implements StartupActivity.RequiredForSmartMode {
  private static final Logger LOG = Logger.getInstance(OnDiskSharedIndexChunkLocator.class);
  private static final String LOCAL_SHARED_INDEX_DIR = "shared-index";
  private static final ExtensionPointName<SharedIndexLocalFinder> LOCAL_FINDER_EP_NAME =
    ExtensionPointName.create("com.intellij.sharedIndexLocalFinderBazel");

  public static final String ROOT_PROP = "on.disk.shared.index.root";

  @Override
  public void runActivity(@NotNull Project project) {
    DumbService.getInstance(project).queueTask(new MyDumbModeTask(project));
  }

  private static void scanAndAttachLocalSharedIndexes(@NotNull Project project,
                                                      @NotNull ProgressIndicator indicator) {
    Path root = getSharedIndexRoot();
    LOG.debug("Scanning " + root + " for manually prepared shared indexes...");

    Set<Path> predefinedPaths = listPredefinedIndexFiles(root);
    Set<Path> customPaths = listCustomIndexFiles(project);

    for (Path ijxPath : ContainerUtil.union(predefinedPaths, customPaths)) {
      ChunkDescriptor chunk = tryLoadCompatibleSharedIndex(ijxPath, SharedIndexInfrastructureVersion.getIdeVersion());
      if (chunk == null) continue;

      try {
        SharedIndexChunkConfiguration.getInstance().downloadChunk(chunk, project, indicator);
      }
      catch (Throwable t) {
        if (t instanceof ControlFlowException) ExceptionUtil.rethrow(t);
        LOG.error("Failed to preload shared index: " + chunk + ". " + t.getMessage(), t);
      }
    }
  }

  @NotNull
  private static Set<Path> listCustomIndexFiles(@NotNull Project project) {
    return LOCAL_FINDER_EP_NAME.extensions().flatMap(finder -> finder.findSharedIndexChunks(project).stream()).collect(Collectors.toSet());
  }

  @NotNull
  private static Set<Path> listPredefinedIndexFiles(@NotNull Path root) {
    if (Files.isRegularFile(root)) {
      return Collections.singleton(root);
    }

    if (Files.isDirectory(root)) {
      try (Stream<Path> pathStream = Files.list(root)) {
        return pathStream
          .filter(f -> f.getFileName().toString().endsWith(".ijx"))
          .collect(Collectors.toSet());
      }
      catch (Exception e) {
        LOG.error("Failed to scan share index file home from " + root + ". " + e.getMessage(), e);
      }
    }

    return Collections.emptySet();
  }

  @NotNull
  private static Path getSharedIndexRoot() {
    String indexRoot = System.getProperty(ROOT_PROP, FileUtil.join(PathManager.getSystemPath(), LOCAL_SHARED_INDEX_DIR));
    return Paths.get(indexRoot);
  }

  private static @Nullable ChunkDescriptor tryLoadCompatibleSharedIndex(@NotNull Path filePath,
                                                                        @NotNull SharedIndexInfrastructureVersion ideVersion) {
    if (!Files.isRegularFile(filePath)) return null;

    LOG.info("Checking local shared index " + filePath);

    @NotNull SharedIndexMetadata version;
    try {
      version = Objects.requireNonNull(getVersion(filePath), "Shared index " + filePath + " contains incompatible metadata");
    }
    catch (Throwable e) {
      LOG.error("Can't fetch shared index version for " + filePath + ". " + e.getMessage(), e);
      return null;
    }

    String fileSize = "(no file size)";
    try {
      fileSize = StringUtil.formatFileSize(Files.size(filePath));
    }
    catch (Exception e) {
      LOG.warn("Can't get size of shared index file " + filePath + ". " + e.getMessage(), e);
    }

    if (!ideVersion.isSuitableMetadata(version)) {
      LOG.warn("Local shared index " + filePath + " is incompatible with current IDE version:\n " +
               "IDE Version: " + ideVersion + "\n" +
               "Index Version: " + version);
      return null;
    }

    LOG.info("Detected local shared index " + filePath + ", " +
             "size " + fileSize + ", " +
             "IDE Version " + ideVersion + ", " +
             "Index Version " + version);

    return new MyChunkDescriptor(filePath, version);
  }

  private static final class MyChunkDescriptor implements ChunkDescriptor {
    private final Path myIndexZip;
    private final @NotNull SharedIndexMetadata myVersion;

    private MyChunkDescriptor(@NotNull Path indexZip, @NotNull SharedIndexMetadata version) {
      myIndexZip = indexZip;
      myVersion = version;
    }

    @Override
    public @NotNull String getChunkUniqueId() {
      return "local-" + myIndexZip.getFileName().toString();
    }

    @NotNull
    @Override
    public String getKind() {
      return myVersion.getIndexKind();
    }

    @Override
    public boolean downloadChunk(@NotNull Path targetFile,
                                 @Nullable Project project,
                                 @NotNull ProgressIndicator indicator) {
      indicator.setText(SharedIndexesBundle.message("configuring.shared.indexes"));
      indicator.setIndeterminate(true);
      LOG.warn("Shared Index " + myIndexZip + " is requested by the IDE");

      try {
        PathKt.copy(myIndexZip, targetFile);
        long size = Files.size(targetFile);

        SharedIndexesFusCollector.INSTANCE.
          reportLocalIndexLoaded(
            project,
            "project", // We don't have kinds for local indexes. Let's consider it is "project"
            getChunkUniqueId(),
            size
          );
        return true;
      }
      catch (Throwable t) {
        if (t instanceof ControlFlowException) ExceptionUtil.rethrow(t);
        LOG.warn("Failed to copy shared index from " + myIndexZip + " to " + targetFile);
        return false;
      }
    }

    @Override
    public String toString() {
      return "Shared Index (" + myVersion + ") from " + myIndexZip;
    }

    @NotNull
    @Override
    public ChunkStorageOption getChunkStorageOption() {
      return ChunkStorageOption.APPEND;
    }
  }

  @Nullable
  private static SharedIndexMetadata getVersion(@NotNull Path indexZip) throws IOException {
    try (UncompressedZipFileSystem fs = UncompressedZipFileSystemProvider.INSTANCE.newFileSystem(indexZip)) {
      return SharedIndexMetadata.readIndexesVersion(fs.getRootDirectory());
    }
  }

  private static class MyDumbModeTask extends DumbModeTask {
    private final @NotNull Project myProject;

    private MyDumbModeTask(@NotNull Project project) {
      super(project);
      myProject = project;
    }

    @Override
    public void performInDumbMode(@NotNull ProgressIndicator indicator) {
      scanAndAttachLocalSharedIndexes(myProject, indicator);
    }

    @Override
    public String toString() {
      return "OnDiskSharedIndexChunkLocator";
    }
  }
}

File: src/main/kotlin/rules_intellij/indexing/OnDiskSharedIndexChunkLocator.kt
// Copyright 2000-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package rules_intellij.indexing

import com.intellij.indexing.shared.message.SharedIndexesBundle
import com.intellij.indexing.shared.metadata.SharedIndexMetadata
import com.intellij.indexing.shared.platform.api.ChunkDescriptor
import com.intellij.indexing.shared.platform.api.SharedIndexInfrastructureVersion
import com.intellij.indexing.shared.platform.impl.ChunkStorageOption
import com.intellij.indexing.shared.platform.impl.SharedIndexChunkConfiguration
import com.intellij.indexing.shared.platform.impl.SharedIndexesFusCollector.reportLocalIndexLoaded
import com.intellij.indexing.shared.util.zipFs.UncompressedZipFileSystemProvider
import com.intellij.openapi.application.PathManager
import com.intellij.openapi.diagnostic.ControlFlowException
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.extensions.ExtensionPointName
import com.intellij.openapi.progress.ProgressIndicator
import com.intellij.openapi.project.DumbModeTask
import com.intellij.openapi.project.DumbService
import com.intellij.openapi.project.Project
import com.intellij.openapi.startup.StartupActivity.RequiredForSmartMode
import com.intellij.openapi.util.io.FileUtil
import com.intellij.openapi.util.text.StringUtil
import com.intellij.util.ExceptionUtil
import com.intellij.util.containers.ContainerUtil
import com.intellij.util.io.copy
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.util.*
import java.util.stream.Collectors

class OnDiskSharedIndexChunkLocator : RequiredForSmartMode {
    override fun runActivity(project: Project) {
        DumbService.getInstance(project).queueTask(MyDumbModeTask(project))
    }

    private class MyChunkDescriptor(private val myIndexZip: Path, private val myVersion: SharedIndexMetadata) : ChunkDescriptor {
        override val chunkUniqueId: String
            get() = "local-" + myIndexZip.fileName.toString()


        override val kind: String
            get() = myVersion.indexKind

        override fun downloadChunk(targetFile: Path,
                                   project: Project?,
                                   indicator: ProgressIndicator): Boolean {
            indicator.text = SharedIndexesBundle.message("configuring.shared.indexes")
            indicator.isIndeterminate = true
            LOG.warn("Shared Index $myIndexZip is requested by the IDE")
            return try {
                myIndexZip.copy(targetFile)
                val size = Files.size(targetFile)
                // We don't have kinds for local indexes. Let's consider it is "project"
                reportLocalIndexLoaded(project, "project", chunkUniqueId, size)
                true
            } catch (t: Throwable) {
                if (t is ControlFlowException) ExceptionUtil.rethrow(t)
                LOG.warn("Failed to copy shared index from $myIndexZip to $targetFile")
                false
            }
        }

        override fun toString(): String {
            return "Shared Index ($myVersion) from $myIndexZip"
        }

        override val chunkStorageOption: ChunkStorageOption
            get() = if (enumContains<ChunkStorageOption>("APPEND"))
                enumValueOf<ChunkStorageOption>("APPEND")
            else
                enumValueOf<ChunkStorageOption>("APPENDABLE")
    }

    private class MyDumbModeTask(private val project: Project) : DumbModeTask(project) {
        override fun performInDumbMode(indicator: ProgressIndicator) {
            scanAndAttachLocalSharedIndexes(project, indicator)
        }

        override fun toString(): String {
            return "OnDiskSharedIndexChunkLocator"
        }
    }

    companion object {
        private val LOG = Logger.getInstance(OnDiskSharedIndexChunkLocator::class.java)
        private const val LOCAL_SHARED_INDEX_DIR = "shared-index"
        private val LOCAL_FINDER_EP_NAME = ExtensionPointName.create<SharedIndexLocalFinder>("com.intellij.sharedIndexLocalFinderBazel")
        const val ROOT_PROP = "on.disk.shared.index.root"
        private fun scanAndAttachLocalSharedIndexes(project: Project, indicator: ProgressIndicator) {
            val root = getSharedIndexRoot()
            LOG.debug("Scanning $root for manually prepared shared indexes...")
            val predefinedPaths = listPredefinedIndexFiles(root)
            val customPaths = listCustomIndexFiles(project)
            for (ijxPath in ContainerUtil.union(predefinedPaths, customPaths)) {
                val chunk = tryLoadCompatibleSharedIndex(ijxPath, SharedIndexInfrastructureVersion.getIdeVersion())
                        ?: continue
                try {
                    SharedIndexChunkConfiguration.getInstance().downloadChunk(chunk, project, indicator)
                } catch (t: Throwable) {
                    if (t is ControlFlowException) ExceptionUtil.rethrow(t)
                    LOG.error("Failed to preload shared index: " + chunk + ". " + t.message, t)
                }
            }
        }

        private fun listCustomIndexFiles(project: Project): Set<Path> {
            return LOCAL_FINDER_EP_NAME.extensions().flatMap { finder: SharedIndexLocalFinder -> finder.findSharedIndexChunks(project).stream() }.collect(Collectors.toSet())
        }

        private fun listPredefinedIndexFiles(root: Path): Set<Path> {
            if (Files.isRegularFile(root)) {
                return setOf(root)
            }
            if (Files.isDirectory(root)) {
                try {
                    Files.list(root).use { pathStream ->
                        return pathStream
                                .filter { f: Path -> f.fileName.toString().endsWith(".ijx") }
                                .collect(Collectors.toSet())
                    }
                } catch (e: Exception) {
                    LOG.error("Failed to scan share index file home from " + root + ". " + e.message, e)
                }
            }
            return emptySet()
        }

        private fun getSharedIndexRoot(): Path {
            val indexRoot = System.getProperty(ROOT_PROP, FileUtil.join(PathManager.getSystemPath(), LOCAL_SHARED_INDEX_DIR))
            return Paths.get(indexRoot)
        }

        private fun tryLoadCompatibleSharedIndex(filePath: Path,
                                                 ideVersion: SharedIndexInfrastructureVersion): ChunkDescriptor? {
            if (!Files.isRegularFile(filePath)) return null
            LOG.info("Checking local shared index $filePath")
            val version = try {
                getVersion(filePath) ?: throw RuntimeException("Shared index $filePath contains incompatible metadata")
            } catch (e: Throwable) {
                LOG.error("Can't fetch shared index version for " + filePath + ". " + e.message, e)
                return null
            }
            var fileSize = "(no file size)"
            try {
                fileSize = StringUtil.formatFileSize(Files.size(filePath))
            } catch (e: Exception) {
                LOG.warn("Can't get size of shared index file " + filePath + ". " + e.message, e)
            }
            if (!ideVersion.isSuitableMetadata(version)) {
                LOG.warn("""Local shared index $filePath is incompatible with current IDE version:
 IDE Version: $ideVersion
Index Version: $version""")
                return null
            }
            LOG.info("Detected local shared index " + filePath + ", " +
                    "size " + fileSize + ", " +
                    "IDE Version " + ideVersion + ", " +
                    "Index Version " + version)
            return MyChunkDescriptor(filePath, version)
        }

        @Throws(IOException::class)
        private fun getVersion(indexZip: Path): SharedIndexMetadata? {
            UncompressedZipFileSystemProvider.INSTANCE.newFileSystem(indexZip).use {
//                fs -> return SharedIndexMetadata.readIndexesVersion(fs.rootDirectory)
                fs -> return SharedIndexMetadata.readSharedIndexMetadata(fs.rootDirectory)
            }
        }
    }
}